File: src/main/java/indi/dkx/laocai/bot/model/segment/Segment.java
package indi.dkx.laocai.bot.model.segment;

import lombok.Data;

import lombok.extern.slf4j.Slf4j;

import tools.jackson.core.JsonParser;
import tools.jackson.databind.DeserializationContext;
import tools.jackson.databind.JsonNode;
import tools.jackson.databind.ValueDeserializer;
import tools.jackson.databind.annotation.JsonDeserialize;
import indi.dkx.laocai.bot.model.segment.data.*;

/**
 * 消息段统一模型。
 * <p>
 * 消息内容在协议里是按 segment 切分的，统一包一层才能在不同类型之间切换解析实现。
 */
@JsonDeserialize(using = Segment.SegmentDeserializer.class)
@Data
public class Segment {
    private final String type;
    private final SegmentData data;
    Segment(String type, SegmentData data) {
        this.type = type;
        this.data = data;
    }

    @Slf4j
    public static class SegmentDeserializer extends ValueDeserializer<Segment> {

        @Override
        public Segment deserialize(JsonParser p, DeserializationContext context) {

            log.debug("进入 Deserializer.deserialize()");

            JsonNode root = p.readValueAsTree();
            String type = root.get("type").asString();
            JsonNode dataNode = root.get("data");
            SegmentData data = switch (type) {
                case "text" -> context.readTreeAsValue(dataNode, TextSegmentData.class);
                case "mention" -> context.readTreeAsValue(dataNode, MentionSegmentData.class);
                case "face" -> context.readTreeAsValue(dataNode, FaceSegmentData.class);
                case "image" -> context.readTreeAsValue(dataNode, IncomingImageSegmentData.class);
                case "reply" -> context.readTreeAsValue(dataNode, IncomingReplySegmentData.class);
                default -> throw new IllegalArgumentException("Unknown type: " + type);
            };
            return new Segment(type, data);
        }
    }
}

File: laocai-bot-spring-boot-starter/src/main/kotlin/indi/kyson/laocai/bot/segment/Segment.kt
package indi.kyson.laocai.bot.segment

import tools.jackson.core.JsonParser
import tools.jackson.databind.DeserializationContext
import tools.jackson.databind.JsonNode
import tools.jackson.databind.ValueDeserializer
import tools.jackson.databind.annotation.JsonDeserialize

/**
 * 消息段统一模型。
 *
 * 消息内容在协议里是按 segment 切分的，每种类型对应一个独立实现类，
 * 类型判断和构造都收敛到各自的类里，而不是依赖外部的 data 类型判断链。
 */
@JsonDeserialize(using = Segment.SegmentDeserializer::class)
sealed interface Segment {

    val type: String

    val data: Any

    /**
     * 反序列化协议返回的入站消息段。
     *
     * 出站消息段由对应的 [OutgoingImageSegment]、[OutgoingReplySegment]、
     * [OutgoingRecordSegment] 工厂方法创建，不在这里反序列化。
     */
    class SegmentDeserializer : ValueDeserializer<Segment>() {

        override fun deserialize(p: JsonParser, context: DeserializationContext): Segment {
            val root: JsonNode = p.readValueAsTree()
            val type = requireNotNull(root.get("type")?.asString()) {
                "Segment type is missing"
            }
            val dataNode = requireNotNull(root.get("data")) {
                "Segment data is missing"
            }
            return when (type) {
                "text" -> context.readTreeAsValue(dataNode, TextSegment.Data::class.java)!!.toSegment()
                "markdown" -> context.readTreeAsValue(dataNode, MarkdownSegment.Data::class.java)!!.toSegment()
                "mention" -> context.readTreeAsValue(dataNode, MentionSegment.Data::class.java)!!.toSegment()
                "face" -> context.readTreeAsValue(dataNode, FaceSegment.Data::class.java)!!.toSegment()
                "image" -> context.readTreeAsValue(dataNode, IncomingImageSegment.Data::class.java)!!.toSegment()
                "reply" -> context.readTreeAsValue(dataNode, IncomingReplySegment.Data::class.java)!!.toSegment()
                else -> UnknownSegment.of(type, dataNode)
            }
        }
    }
}
